src/main.rs

```diff
@@ -311,38 +299,25 @@ fn workspace(config: &Config, manifest_path: Option<PathBuf>) -> CargoResult<Wor
 
 fn registry<'a>(config: &'a Config, package: &Package) -> CargoResult<PackageRegistry<'a>> {
     let mut registry = PackageRegistry::new(config)?;
-    registry.add_sources(Some(package.package_id().source_id().clone()))?;
+    registry.add_sources(Some(package.package_id().source_id()))?;
     Ok(registry)
 }
 
 fn resolve<'a, 'cfg>(
     registry: &mut PackageRegistry<'cfg>,
     workspace: &'a Workspace<'cfg>,
     features: Option<String>,
-    all_features: bool,
     no_default_features: bool,
     no_dev_dependencies: bool,
 ) -> CargoResult<(PackageSet<'a>, Resolve)> {
-    let features = Method::split_features(&features.into_iter().collect::<Vec<_>>());
+    let features: Vec<String> = features.into_iter().collect();
 
     let (packages, resolve) = ops::resolve_ws(workspace)?;
 
-    let method = Method::Required {
-        dev_deps: !no_dev_dependencies,
-        features: Rc::new(features),
-        all_features,
-        uses_default_features: !no_default_features,
-    };
+    let opts = ResolveOpts::new(!no_dev_dependencies, &features, true, !no_default_features);
 
-    let resolve = ops::resolve_with_previous(
-        registry,
-        workspace,
-        method,
-        Some(&resolve),
-        None,
-        &[],
-        true,
-    )?;
+    let resolve =
+        ops::resolve_with_previous(registry, workspace, opts, Some(&resolve), None, &[], true)?;
```

Comment:
How can I work around all_features things while the 6th parameter of `resolve_with_previous`
is always `&[]`, which means all_features passed to `ResolveOpts::new` must be true?

Reply:
I don't know off the top of my head. You could check the cargo source to see how its commands implement `--all-features`.

Reply:
I read cargo's source in until current master 4647b1d142a81de3bbd6be36105fa959a9c01f8f.
All uses of resolve_with_previous is with `ResolveOpts::everything()`, which has all_features=true.
We could keep the all-feature cmd flag but it may not do anything useful.

Reply:
Are you claiming that the behavior of `--all-features` in Cargo's CLI interface has changed to not do anything useful?

Reply:
No. I meant that the use of all_features in resolve_with_previous when specs is empty is useless.
All other uses of all-features in cargo is with non-empty specs.

Reply:
Okay, so it sounds like this PR would need to pass in non-empty specs then, right?

Reply:
Yeah, but what do we get that specs from?

Reply:
I don't know, but the cargo source would probably be the place to look.

Reply:
Done.